beluga_amcl/src/amcl_node.cpp

```diff
@@ -816,19 +818,67 @@ void AmclNode::timer_callback() {
     return;
   }
 
+  // Particle weights from the filter may or may not be representative of the
+  // true distribution. If we resampled, they are not, and there will be multiple copies
+  // of the most likely candidates, all with unit weight. In this case the number of copies
```

Comment:
@glpuga meta: it makes me wonder, are we copying more than we should? Resampling over particle indices (whatever those indices may be) to build up weights instead of duplicating particles wouldn't be particularly hard. Not sure if sequential copying trumps random access because of caching though.

Reply:
The benefits would also not last long, since the next sampling would already force all the particles to become different and make indexing an overhead.

Also we sample far more frequently than we resample; having multiple identical copies is the exemption more than the rule. It just happens that we basically only publish this message when we resample (assuming the usual resample_interval == 1 case), so dealing with that situation is important for this function, but not so much from a performance perspective.